### PR TITLE
feat(state_table): add iterator sub range under a certain pk prefix

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
https://github.com/risingwavelabs/risingwave/issues/11891

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by hzxa21 — COMMENTED** [review]
I am thinking whether we can unify `iter_row_with_pk_prefix_sub_range` and `iter_row_with_pk_range` into one or at least we can simplfy the implementation since they have many things in common. One question needs to be clarified before discussing whether the unification is possible:
- Is the `pk_prefix` passed to `iter_row_with_pk_prefix_sub_range` intended to be used as a prefix hint (i.e. bloom filter key) for state store?

**Comment by wcy-fdu** [comment]
> I am thinking whether we can unify `iter_row_with_pk_prefix_sub_range` and `iter_row_with_pk_range` into one or at least we can simplfy the implementation since they have many things in common. One question needs to be clarified before discussing whether the unification is possible:
> 
> * Is the `pk_prefix` passed to `iter_row_with_pk_prefix_sub_range` intended to be used as a prefix hint (i.e. bloom filter key) for state store?

After discussion, we reached the following agreement:
- StateTable provide two types of iterator interfaces to the upper layer, one is that the executors know `pk_prefix`, the other is that it does not know. The former does not need to pass vnode, the latter requires.
- Replace the `iter_row_with_pk_prefix` interface with the new `iter_row_with_pk_prefix_sub_range`, as `iter_with_pk_prefix` is a special case where sub_range in `iter_row_with_pk_prefix_sub_range` is` (unbounded, unbounded).` (This will be done in next PR)
- `iter_row_with_pk_prefix_sub_range` and `iter_row_with_pk_range` is both needed, but within the state table their implementations can be merged.

File: src/stream/src/common/table/state_table.rs [hunk 1]
```diff
@@ -1195,6 +1195,30 @@ where
...
+        sub_range: &(Bound<impl Row>, Bound<impl Row>),
+        prefetch_options: PrefetchOptions,
+    ) -> StreamExecutorResult<KeyedRowStream<'_, S, SD>> {
+        let vnode = self.compute_prefix_vnode(&pk_prefix).to_be_bytes();
+
+        let memcomparable_range = prefix_and_sub_range_to_memcomparable(
+            &self.pk_serde,
+            sub_range,
+            pk_prefix.to_owned_row(),
+        );
```
Comment (stdrc) [inline src/stream/src/common/table/state_table.rs:1212]:
```suggestion
        let memcomparable_range =
            prefix_and_sub_range_to_memcomparable(&self.pk_serde, sub_range, pk_prefix);
```

File: src/stream/src/common/table/state_table.rs [hunk 2]
```diff
@@ -1297,15 +1321,50 @@ pub fn prefix_range_to_memcomparable(
...
+        Excluded(end_range) => {
+            Bound::Excluded((pk_prefix_calculate_end_range.chain(end_range)).to_owned_row())
+        }
+        Unbounded => Unbounded,
+    };
+    (
+        start_range_to_memcomparable(pk_serde, &start_range),
+        end_range_to_memcomparable(pk_serde, &end_range, Some(serialized_pk_prefix)),
+    )
+}
```
Comment (stdrc) [inline src/stream/src/common/table/state_table.rs:1363]:
```suggestion
fn prefix_and_sub_range_to_memcomparable(
    pk_serde: &OrderedRowSerde,
    sub_range: &(Bound<impl Row>, Bound<impl Row>),
    pk_prefix: impl Row,
) -> (Bound<Bytes>, Bound<Bytes>) {
    let (range_start, range_end) = sub_range;
    let prefix_serializer = pk_serde.prefix(pk_prefix.len());
    let serialized_pk_prefix = serialize_pk(&pk_prefix, &prefix_serializer);

    let start_range = match range_start {
        Included(start_range) => Bound::Included(Either::Left((&pk_prefix).chain(start_range))),
        Excluded(start_range) => Bound::Excluded(Either::Left((&pk_prefix).chain(start_range))),
        Unbounded => Bound::Included(Either::Right(&pk_prefix)),
    };

    let end_range = match range_end {
        Included(end_range) => Bound::Included((&pk_prefix).chain(end_range)),
        Excluded(end_range) => Bound::Excluded((&pk_prefix).chain(end_range)),
        Unbounded => Unbounded,
    };
    (
        start_range_to_memcomparable(pk_serde, &start_range),
        end_range_to_memcomparable(pk_serde, &end_range, Some(serialized_pk_prefix)),
    )
}
```

Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:1363]: Use `Either` to avoid `into_owned_row`

Comment (wcy-fdu) [reply src/stream/src/common/table/state_table.rs]: Thanks.

**Review by stdrc — APPROVED** [review]
LGTM

**Review by st1page — APPROVED** [review]

**Pull request merged**